lib/vm/src/instance/ref.rs

```diff
@@ -109,114 +52,102 @@ impl InstanceRef {
-            return;
-        }
-
-        // This fence is needed to prevent reordering of use of the data and
-        // deletion of the data. Because it is marked `Release`, the decreasing
-        // of the reference count synchronizes with this `Acquire` fence. This
-        // means that use of the data happens before decreasing the reference
-        // count, which happens before this fence, which happens before the
-        // deletion of the data.
-        //
-        // As explained in the [Boost documentation][1]:
-        //
-        // > It is important to enforce any possible access to the object in one
-        // > thread (through an existing reference) to *happen before* deleting
-        // > the object in a different thread. This is achieved by a "release"
-        // > operation after dropping a reference (any access to the object
-        // > through this reference must obviously happened before), and an
-        // > "acquire" operation before deleting the object.
-        //
-        // [1]: https://www.boost.org/doc/libs/1_55_0/doc/html/atomic/usage_examples.html
-        atomic::fence(atomic::Ordering::Acquire);
-
-        // Now we can deallocate the instance. Note that we don't
-        // check the pointer to `Instance` is correctly initialized,
-        // but the way `InstanceHandle` creates the
-        // `InstanceRef` ensures that.
         unsafe { Self::deallocate_instance(self) };
     }
 }
 
-impl MemoryUsage for InstanceRef {
+/// TODO: Review this super carefully.
+unsafe impl Send for InstanceInner {}
+unsafe impl Sync for InstanceInner {}
+
+impl MemoryUsage for InstanceInner {
     fn size_of_val(&self, tracker: &mut dyn MemoryUsageTracker) -> usize {
-        mem::size_of_val(self) + self.strong.size_of_val(tracker) - mem::size_of_val(&self.strong)
-            + self.instance_layout.size_of_val(tracker)
+        mem::size_of_val(self) + self.instance_layout.size_of_val(tracker)
             - mem::size_of_val(&self.instance_layout)
             + self.as_ref().size_of_val(tracker)
             - mem::size_of_val(&self.instance)
     }
 }
+
+/// An `InstanceRef` is responsible to properly deallocate,
+/// and to give access to an `Instance`, in such a way that `Instance`
+/// is unique, can be shared, safely, across threads, without
+/// duplicating the pointer in multiple locations. `InstanceRef`
+/// must be the only “owner” of an `Instance`.
+///
+/// Consequently, one must not share `Instance` but
+/// `InstanceRef`. It acts like an Atomically Reference Counter
+/// to `Instance`. In short, `InstanceRef` is roughly a
+/// simplified version of `std::sync::Arc`.
+///
+/// This `InstanceRef` must be freed with [`InstanceRef::deallocate_instance`]
+/// if and only if it has been set correctly. The `Drop` implementation of
+/// [`InstanceRef`] calls its `deallocate_instance` method without
+/// checking if this  property holds, only when `Self.strong` is equal to 1.
```

Comment:
This part of the comment must be updated please :-)!

Reply:
Thanks!  For sure on the bottom part, I actually thought about removing the 

> In short, `InstanceRef` is roughly a simplified version of `std::sync::Arc`.</span></span>

part because it's not true in implementation anymore but it's still true in API, which is arguably more important for the doc comment to express. `InstanceRef`'s methods are pretty much just a subset of `Arc`'s methods.